### ISSUE TITLE
Pull members up to base type breaks existing usages if the method is private

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#36786]
VSF_TYPE_MARKDOWNHi,

The &quot;Pull members up to base type&quot; refactor, when the method being pulled is private, will keep its visibility and will break existing usages on the current derived type.

My expectation would be to use this refactor when I create a method in a derived type and want all derived types of a certain base type to be able to use it as well. If the method is not changed to &quot;protected&quot; instead of private, that doesn't happen and all the usages of that function in the current class are broken.

To repro:
1. Create a base class A and a derived class B.
2. Create a private method in class B and use it from somewhere inside that class.
3. Use the &quot;Pull members up to base type&quot; refactor to pull that private method up to A.
4. All usages are broken.

I'm reporting this because I would not see any other possible use for this refactor in a private method than being able to access it from all classes derived from the base, and right now I am forced to go onto the base class and manually change the visibility of the method.

Thanks,
David

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/599195/pull-members-up-to-base-type-breaks-existing-usage.html
VSTS ticketId: 911340_
_These are the original issue comments:_

Visual Studio Feedback System on 6/10/2019, 01:36 AM (16 days ago): <p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

Visual Studio Feedback System on 6/26/2019, 09:53 AM (51 min ago): <markdown><p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>
</markdown>
_These are the original issue solutions:_
(no solutions)

**Comment by Cosifne** [comment]
Some thoughts:
1. Is it true that people really don't want to move a private method up? I think probably we don't want to lose this ability.
1. If we want to fulfill his request, public, internal and internal protected are also feasible modifiers. How to fulfill those? (Maybe via giving a drop-down list?) Would this become providing all modifier combinations?
So a more general question would be: what's the baseline of refactoring? Should we give every possible option to the user and do everything for him/her? (For this case, I think the user's cost to fix it is quite cheap, just go to the method and change one word would be enough) Or at least we shouldn't break the syntax?

**Comment by dave96** [comment]
Hi,

I wrote that original feedback so I'll try to explain.

1. I don't see how pulling a private member up makes sense, but maybe there's a use case I've never had. However, my real issue is with pulling private members _breaking_ the current usages. I think the main point of refactors is that code keeps working afterwards, as happens with every single other refactor in VS2019.

2. Pulling public and protected members up is not a problem, since it doesn't break usages. Pulling internal methods _can_ be a problem if the class is in another Assembly, so I do think if the current behaviour is to keep them, we should at least drop the internal when crossing Assemblies.

On a more personal note, I don't think refactoring is about choosing between a thousand dropdown menus, because that ends up not being that time-saving. Refactoring is about saving time on common refactors you would do as a programmer (not every single one, just the common!) and there is something that happens after every single refactor I do: The code keeps working :)

Hopefully I managed to explain why I thought of this as an issue and why I think it's important to fix it.

**Comment by CyrusNajmabadi** [comment]
>  I think the main point of refactors is that code keeps working afterwards, as happens with every single other refactor in VS2019.

This is not the case.  I did a survey one time, and i found that all (or all but one) refactoring could lead to broken code.

Our philosophy on refactorings is that they be generally available to help speed you up.  This was based on non-stop feedback in the early years of people saying "the refactoring isn't showing up, so i can't trust that i can use it, so i don't use it".   It turned out that *a lot of the time* code either is in error, or people want to do move it through intermediate points in time that *will be in error*.  Not allowing them to do this is just limiting and fights against very real work pattern that people like getting help with.

**Comment by dave96** [comment]
> This is not the case. I did a survey one time, and i found that all (or all but one) refactoring could lead to broken code.

Well this is not what I have experienced so far, but I understand I'm not the only use case that exists :). However, I am making this argument because I do believe it can be useful for other users.

Your point is refactoring is meant to speed me up. Right, I agree as well with that. One of my points in the original text, is that this refactor on its current form **cannot** speed me up in the case of private methods. Let's see the two use cases:

1. **Current**: I want to move a private method to my base class. In which case I use the refactor, and then I _have to go open my base class anyway to use it_. At that point, I might as well have cut-n-pasted it.
2. **Proposed**: What I actually want is to make that method available to all inheritants of my base class, because I implemented some code that might be useful for them. So I use this refactor, nothing breaks, and I can go to my other classes and use it. _I did not have to open the base class at all, so this is faster than cut-n-paste_.

Right now, 1 is no faster that cut-n-paste, and 2 is as slow as 1 because I have to open my base class to change the visibility. So in my point of view, we can only win with this change :). Ain't that nice?

**Comment by CyrusNajmabadi** [comment]
> Right now, 1 is no faster that cut-n-paste

I disagree.  I will often pull up many members at once including private members that are needed by other members i'm pulling up (or just private members that i want to move, knowing that i'll have to fixup some stuff afterwards).  Being able to move everything up in one step is a lot faster than N cut-and-paste steps.

**Comment by petrroll** [comment]
Just my 2-cents: Moving private methods make sense because (I) usually try to have the most restrictive accessor. And so if it's the leaf class, it will have methods private. If I decide to pull them up, I want them to become `protected`.

As for being safe, can't we simply correct it & show a warning (& handle it as all potentially breaking refactorings)?
